flink/src/main/java/org/apache/iceberg/flink/source/FlinkSource.java

```diff
@@ -89,8 +88,18 @@ public Builder table(Table newTable) {
       return this;
     }
 
-    public Builder filters(List<Expression> newFilters) {
-      this.filterExpressions = newFilters;
+    public Builder hadoopConf(Configuration newConf) {
```

Comment:
We try to avoid using Hadoop classes in Iceberg APIs because they are hard to remove. Injecting a Hadoop `Configuration` is currently done in one place: to instantiate a catalog that requires it.

The catalog creates tables and tables are associated with a `FileIO`, so the configuration is passed down through that chain. The table's configuration should be used for any table configuration.

MR also has a Hadoop `Configuration`, but that's required by the API so we can't remove it from the API there. But we still prefer using the table's `Configuration` when it is needed for components like `HadoopFileIO`.

Reply:
+1 for avoiding using Hadoop `Configuration`.

Reply:
The reason why Hadoop conf needs to be passed now is because:
- JobManager needs the splits of the scan, so it needs to call `Table.newScan`.
- So JobManager needs `Table` object, where can a table be generated? `TableLoader` -> from `Catalog` or `HadoopTables`.
- The creation of `Catalog` and `HadoopTables` needs a Hadoop `Configuration`.

Maybe we can pass this chain with an Iceberg object (`FileIO` may looks not good to the creation of catalog)?

Reply:
Maybe `CatalogLoader` should serialize its own `Configuration`? It would make sense to pass one when creating a `CatalogLoader` for `HadoopCatalog` or `HiveCatalog` because those need a `Configuration` to create the catalog.

Reply:
Or maybe this could use the approach from the `FlinkCatalogFactory`:

```java
  public static Configuration clusterHadoopConf() {
    return HadoopUtils.getHadoopConfiguration(GlobalConfiguration.loadConfiguration());
  }
```

Using `clusterHadoopConf` internally here would avoid the need to expose this in the API and we could add `Configuration` to the loader later.

Reply:
Using `clusterHadoopConf` may not so flexible.
I am +1 for CatalogLoader should serialize its own Configuration. I will create a PR later for source and sink.